### ISSUE TITLE
`pip install stellargraph` results in a broken stellargraph: GCN fails to run with "TypeError: Using a `tf.Tensor` as a Python `bool` is not allowed"

### ISSUE THREAD

**Issue by huonw** [stellargraph/stellargraph#488]
**Workaround**

Downgrade keras:

- manually: `pip install keras<2.2.5`
- in a requirements file (or similar): add `keras<2.2.5`

**Describe the bug**

Keras 2.3.0 was released today (https://pypi.org/project/Keras/#history, https://github.com/keras-team/keras/releases/tag/2.3.0), and that release causes causes the following error when running GCN:

```
Using GCN (local pooling) filters...
Traceback (most recent call last):
  File "<stdin>", line 253, in <module>
  File "<stdin>", line 179, in main
  File "<stdin>", line 71, in train
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 418, in node_model
    x_out = self(x_inp)
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 377, in __call__
    h_layer = layer([h_layer, out_indices] + Ainput)
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/keras/engine/base_layer.py", line 489, in __call__
    output = self.call(inputs, **kwargs)
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 210, in call
    if self.bias:
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 777, in __bool__
    return bool(self.read_value())
  File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 690, in __bool__
    raise TypeError("Using a `tf.Tensor` as a Python `bool` is not allowed. "
TypeError: Using a `tf.Tensor` as a Python `bool` is not allowed. Use `if t is not None:` instead of `if t:` to test if a tensor is defined, and use TensorFlow ops such as tf.cond to execute subgraphs conditioned on the value of a tensor.
```

**To Reproduce**
1. Create a clean virtualenv and switch to it (e.g. `pyenv virtualenv sg-keras-test && pyenv activate sg-keras-test`
2. Run `pip install stellargraph` (note `keras-2.3.0` `stellargraph-0.7.1` `tensorflow-1.14.0`)
   ```
   ...

   Successfully installed absl-py-0.8.0 astor-0.8.0 boto-2.49.0 boto3-1.9.230 
     botocore-1.12.230 certifi-2019.9.11 chardet-3.0.4 cycler-0.10.0 decorator-4.4.0 
     docutils-0.15.2 gast-0.3.2 gensim-3.8.0 google-pasta-0.1.7 grpcio-1.23.0 
     h5py-2.10.0 idna-2.8 jmespath-0.9.4 joblib-0.13.2 keras-2.3.0 
     keras-applications-1.0.8 keras-preprocessing-1.1.0 kiwisolver-1.1.0 
     markdown-3.1.1 matplotlib-3.1.1 networkx-2.3 numpy-1.17.2 pandas-0.25.1 
     protobuf-3.9.1 pyparsing-2.4.2 python-dateutil-2.8.0 pytz-2019.2 pyyaml-5.1.2 
     requests-2.22.0 s3transfer-0.2.1 scikit-learn-0.21.3 scipy-1.3.1 six-1.12.0 
     smart-open-1.8.4 stellargraph-0.7.1 tensorboard-1.14.0 tensorflow-1.14.0 
     tensorflow-estimator-1.14.0 termcolor-1.1.0 urllib3-1.25.3 werkzeug-0.15.6 
     wheel-0.33.6 wrapt-1.11.2
   ```
3. Run the stellargraph 0.7.1 gcn example, e.g. if the dataset is in `~/data/Cora`
   ```
   curl https://raw.githubusercontent.com/stellargraph/stellargraph/v0.7.1/demos/node-classification/gcn/gcn-cora-example.py | python - -l ~/data/Cora/
   ```
   and observe the failure above. Output with more context:
   ```
   Using TensorFlow backend.
   /Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
     _np_qint8 = np.dtype([("qint8", np.int8, 1)])

   ...

   /Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
     np_resource = np.dtype([("resource", np.ubyte, 1)])
   Using GCN (local pooling) filters...
   Traceback (most recent call last):
     File "<stdin>", line 253, in <module>
     File "<stdin>", line 179, in main
     File "<stdin>", line 71, in train
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 418, in node_model
       x_out = self(x_inp)
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 377, in __call__
       h_layer = layer([h_layer, out_indices] + Ainput)
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/keras/engine/base_layer.py", line 489, in __call__
       output = self.call(inputs, **kwargs)
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/stellargraph/layer/gcn.py", line 210, in call
       if self.bias:
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/ops/resource_variable_ops.py", line 777, in __bool__
       return bool(self.read_value())
     File "/Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 690, in __bool__
       raise TypeError("Using a `tf.Tensor` as a Python `bool` is not allowed. "
   TypeError: Using a `tf.Tensor` as a Python `bool` is not allowed. Use `if t is not None:` instead of `if t:` to test if a tensor is defined, and use TensorFlow ops such as tf.cond to execute subgraphs conditioned on the value of a tensor.
   ```
4. To confirm that it's keras 2.3, downgrade keras: `pip install keras<2.3`
   ```
   Installing collected packages: keras
     Found existing installation: Keras 2.3.0
       Uninstalling Keras-2.3.0:
         Successfully uninstalled Keras-2.3.0
   Successfully installed keras-2.2.5
   ```
5. Rerun step 3, and observe that it is successful:
   ```
   Using TensorFlow backend.
   /Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
     _np_qint8 = np.dtype([("qint8", np.int8, 1)])

   ...

   WARNING:tensorflow:From /Users/wil9dw/.pyenv/versions/sg/lib/python3.6/site-packages/tensorflow/python/ops/math_grad.py:1250: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
   Instructions for updating:
   Use tf.where in 2.0, which has the same broadcast rule as np.where
   Epoch 1/50
   2019-09-18 09:48:50.830160: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
    - 0s - loss: 1.9754 - weighted_categorical_accuracy: 0.2286 - val_loss: 1.9001 - val_weighted_categorical_accuracy: 0.3200

   ...

   Epoch 50/50
    - 0s - loss: 0.2165 - weighted_categorical_accuracy: 1.0000 - val_loss: 0.7909 - val_weighted_categorical_accuracy: 0.8067
   Model: "model_1"
   __________________________________________________________________________________________________
   Layer (type)                    Output Shape         Param #     Connected to                     
   ==================================================================================================
   input_1 (InputLayer)            (1, 2708, 1433)      0                                            
   __________________________________________________________________________________________________
   input_3 (InputLayer)            (1, None, 2)         0                                            
   __________________________________________________________________________________________________
   input_4 (InputLayer)            (1, None)            0                                            
   __________________________________________________________________________________________________
   dropout_1 (Dropout)             (1, 2708, 1433)      0           input_1[0][0]                    
   __________________________________________________________________________________________________
   input_2 (InputLayer)            (1, None)            0                                            
   __________________________________________________________________________________________________
   squeezed_sparse_conversion_2 (S (2708, 2708)         0           input_3[0][0]                    
                                                                    input_4[0][0]                    
   __________________________________________________________________________________________________
   graph_convolution_1 (GraphConvo (1, 2708, 32)        45888       dropout_1[0][0]                  
                                                                    input_2[0][0]                    
                                                                    squeezed_sparse_conversion_2[0][0
   __________________________________________________________________________________________________
   dropout_2 (Dropout)             (1, 2708, 32)        0           graph_convolution_1[0][0]        
   __________________________________________________________________________________________________
   graph_convolution_2 (GraphConvo (1, None, 7)         231         dropout_2[0][0]                  
                                                                    input_2[0][0]                    
                                                                    squeezed_sparse_conversion_2[0][0
   ==================================================================================================
   Total params: 46,119
   Trainable params: 46,119
   Non-trainable params: 0
   __________________________________________________________________________________________________
   None
   
   Test Set Metrics:
   	loss: 0.8285
   	weighted_categorical_accuracy: 0.7976
   ```

**Expected behavior**

`pip install stellargraph` should install a working module, by default.

**Desktop (please complete the following information):**
 - OS: macOS
 - Version 10.14.5 (18F203)

**Data used**
Cora

**Additional context**

The particular line of code that's demonstrating this problem is already fixed on `develop` (by #478): https://github.com/stellargraph/stellargraph/blob/a1324a8115d1f148f20a718a2245d84f32440e08/stellargraph/layer/gcn.py#L212

To confirm, follow the instructions above except, in step 2, `pip install git+https://github.com/stellargraph/stellargraph.git@develop` and, in step 3, change the URL to run the `develop` example (not `v0.7.1`):
```
curl https://raw.githubusercontent.com/stellargraph/stellargraph/develop/demos/node-classification/gcn/gcn-cora-example.py | python - -l ~/data/Cora/
```

Before that PR landed, the keras version on `develop` was also restricted to be less than 2.2.5: https://github.com/stellargraph/stellargraph/commit/3817ec5ffb9e7601481ce73f48af51bd4e0ac041

**Comment by youph** [comment]
A hotfix `0.7.2` that fixes this is released.

**Comment by huonw** [comment]
Following the reproducing instructions above no longer reproduces the issues 🎉 

For instance:

```
$ pip install stellargraph
Collecting stellargraph
...
Successfully installed absl-py-0.8.0 astor-0.8.0 boto-2.49.0 boto3-1.9.232 botocore-1.12.232 
  certifi-2019.9.11 chardet-3.0.4 cycler-0.10.0 decorator-4.4.0 docutils-0.15.2 gast-0.3.2 
  gensim-3.8.0 google-pasta-0.1.7 grpcio-1.23.0 h5py-2.10.0 idna-2.8 jmespath-0.9.4 joblib-0.13.2 
  keras-2.2.4 keras-applications-1.0.8 keras-preprocessing-1.1.0 kiwisolver-1.1.0 markdown-3.1.1 
  matplotlib-3.1.1 networkx-2.3 numpy-1.17.2 pandas-0.25.1 protobuf-3.9.1 pyparsing-2.4.2 
  python-dateutil-2.8.0 pytz-2019.2 pyyaml-5.1.2 requests-2.22.0 s3transfer-0.2.1 scikit-learn-0.21.3 
  scipy-1.3.1 six-1.12.0 smart-open-1.8.4 stellargraph-0.7.2 tensorboard-1.14.0 tensorflow-1.14.0 
  tensorflow-estimator-1.14.0 termcolor-1.1.0 urllib3-1.25.5 werkzeug-0.16.0 wheel-0.33.6 wrapt-1.11.2
```

and the example runs successfully.

**Comment by youph** [comment]
The issue has been fixed in release 0.7.2 (hotfix), closing.
